pkg/resourcesets/collector.go

```diff
@@ -248,157 +174,147 @@ func (h *ResourceHandler) filterByLabel(ctx context.Context, dr dynamic.Resource
 	return resourceObjectsList, nil
 }
 
+func (h *ResourceHandler) filterByKind(filter v1.ResourceSelector, apiResources []k8sv1.APIResource) ([]k8sv1.APIResource, error) {
+	var resourceList []k8sv1.APIResource
+	var kindRegexp *regexp.Regexp
+	if filter.KindsRegexp == "" && len(filter.Kinds) == 0 {
+		// if no filters for resource kind are given, return entire resource list
+		return apiResources, nil
+	}
+	if filter.KindsRegexp != "" {
+		var err error
+		kindRegexp, err = regexp.Compile(filter.KindsRegexp)
+		if err != nil {
+			return resourceList, err
+		}
+	}
+	allowedNames := make(map[string]bool)
+	disallowedNames := make(map[string]bool)
+	for _, name := range filter.Kinds {
+		allowedNames[name] = true
+	}
+	for _, name := range filter.ExcludeKinds {
+		disallowedNames[name] = true
+	}
+
+	// "resources" list has all resources under given groupVersion, first filter based on KindsRegexp
+	// Look for a match in either `Kind` (singular name) or `Name` (plural name).
+	// If an exclusion includes either of `Kind` or `Name`, then it's excluded, even if we matched on the other.
+	for _, resObj := range apiResources {
+		includeIt := allowedNames[resObj.Kind] || allowedNames[resObj.Name]
+		if !includeIt && kindRegexp != nil {
```

Comment:
Not sure I'm following all the filtering logic here, it seems to be doing more than I would expect. Unless I'm wrong I think our psuedo-logic here is:

```
IncludeObject = (Match Kinds || !Match ExcludeKinds && Match KindsRegex)
```

Since I think it would be illogical to support users creating this:

```
  - apiVersion: apiextensions.k8s.io/v1
    kinds:
      - clusters
      - pods
    excludeKinds:
      - pods
```

But I think , this should be supported still:
```
  - apiVersion: apiextensions.k8s.io/v1
    kindsRegexp: "^c"
    kinds:
      - pods
    excludeKinds:
      - clusterroles
```